Split the 'lib/happychat/connection' module into a separate chunk and lazy-load it

jsnajdr opened the pull request:
The HappyChat client uses the socket.io library, which is rather big. This patch separates the `lib/happychat/connection` library into separate Webpack chunk and async loads it only when really needed, i.e., first time any method on `Connection` is called.

The resulting `build` chunk reduction is very nice:
```
chunk                                     stat_size           parsed_size           gzip_size
async-load-my-sites-stats-site                +82 B  (+0.0%)        +87 B  (+0.0%)      +22 B  (+0.1%)
async-load-my-sites-stats-stats-insights      +82 B  (+0.0%)       +101 B  (+0.1%)      +18 B  (+0.0%)
async-load-my-sites-stats-summary             +82 B  (+0.0%)       +101 B  (+0.1%)      +17 B  (+0.1%)
async-load-reader-following-manage          +2280 B  (+0.5%)       +952 B  (+0.5%)     +295 B  (+0.7%)
build                                     -172616 B  (-4.6%)     -68677 B  (-4.3%)   -19519 B  (-5.0%)
devdocs                                     +2280 B  (+0.1%)       +952 B  (+0.0%)     +298 B  (+0.1%)
manifest                                       +0 B                +137 B  (+1.3%)      +26 B  (+0.8%)
security                                      +82 B  (+0.0%)       +101 B  (+0.0%)      +23 B  (+0.0%)
settings-traffic                              +82 B  (+0.0%)       +101 B  (+0.0%)      +19 B  (+0.0%)
async-load-lib-happychat-connection       +174118 B    (new)     +69623 B    (new)   +20369 B    (new)
```

The new chunk is 68kB parsed and 20kB gzipped big. Quite a lot of NPM packages there, needed only by socket.io and nothing else:

<img width="752" alt="happychat-connection-chunk" src="https://user-images.githubusercontent.com/664258/35622882-bd63967e-0689-11e8-964e-ee47c303e5d7.png">

**How to test:**
Test that HappyChat still works without any regressions. Establish a chat, check that you can send and receive messages, that the HE on the other side sees all the information they are supposed to see, that the chat is re-established after page reload... There are surely other HappyChat workflows I'm not aware of -- @nosolosw please could you have a look with your team?

nosolosw commented:
I've tested a few flows that use the three connection methods (`init`, `send`, and `request`) and they work as expected. 

Code-wise, the only suggestion I have is to move the new code to `lib/happychat/connection-async.js` and import the `buildConnection` function from the middleware. The core rationale for this request is that I'd like this middleware to remain as much similar as possible to [this other one](https://github.com/Automattic/happychat-client/blob/master/src/state/middleware.js) (in the happychat-client project). All Happychat Calypso code will be removed eventually to use the happychat-client library instead, and the fewer differences between the two codebases, the easier it'll be to tackle that. I hope you may make this change but I'm giving you a :ship: so you can move at full speed without further async (pun intended) back-and-forths.

nosolosw commented on client/state/happychat/middleware.js:
```diff
@@ -30,6 +29,38 @@ const eventMessage = {
 	HAPPYCHAT_FOCUS: 'Started looking at Happychat',
 };
 
+// This function creates a stub `Connection` object stub that dynamically loads the chunk
```
Should this read `This function creates a stub `Connection` object that dynamically loads the chunk` instead?

jsnajdr replied:
Yes, fixed 👍

jsnajdr commented:
@nosolosw I implemented your suggestion and moved the `buildConnection` function to `lib/happychat/connection-async.js`. The only change in `happychat/middleware.js` is now the import.
